typing/env.ml

```
@@ -1965,33 +1999,40 @@ and store_extension ~check ~rebind id addr ext env =
     constrs = TycompTbl.add id cda env.constrs;
     summary = Env_extension(env.summary, id, ext) }
 
-and store_module ~check ~freshening_sub id addr presence md env =
-  let loc = md.md_loc in
+and store_module ?(update_summary=true) ~check ~freshening_sub id addr presence md env =
+  let open Subst.Lazy in
+  let loc = md.mdl_loc in
   Option.iter
-    (fun f -> check_usage loc id md.md_uid f !module_declarations) check;
-  let alerts = Builtin_attributes.alerts_of_attrs md.md_attributes in
+    (fun f -> check_usage loc id md.mdl_uid f !module_declarations) check;
+  let alerts = Builtin_attributes.alerts_of_attrs md.mdl_attributes in
   let module_decl_lazy =
     match freshening_sub with
-    | None -> Lazy_backtrack.create_forced md
-    | Some s -> Lazy_backtrack.create (s, Subst.Rescope (Ident.scope id), md)
+    | None -> md
+    | Some s -> module_decl (Rescope (Ident.scope id)) s md
   in
   let comps =
-    components_of_module ~alerts ~uid:md.md_uid
-      env freshening_sub Subst.identity (Pident id) addr md.md_type
+    components_of_module ~alerts ~uid:md.mdl_uid
+      env freshening_sub Subst.identity (Pident id) addr md.mdl_type
   in
   let mda =
     { mda_declaration = module_decl_lazy;
       mda_components = comps;
       mda_address = addr }
   in
+  let summary =
+    if not update_summary then env.summary
```

Rather than these `update_summary` parameters, perhaps it would be better to be able to disable the summary in an environment, and then nothing would get put there. That seems less likely to allow us to silently drop things from summaries by accident because any attempt to extract the disabled summary can raise an assertion failure.

---

I'm going to leave this as is for the moment, to avoid making the patch any bigger. I think the "right" solution is to split the env type into two, so that we can use envs that explicitly, statically lack a summary in those contexts where the summary is not needed.